traitlets/config/application.py

```diff
@@ -962,6 +963,29 @@ def load_config_file(self, filename, path=None):
         new_config.merge(self.cli_config)
         self.update_config(new_config)
 
+    @catch_config_error
+    def load_config_environ(self):
+        """Load config files by environment."""
+
+        PREFIX = self.name.upper()
+        new_config = Config()
+
+        self.log.debug('Looping through config variables with prefix "%s"', PREFIX)
+
+        for k, v in os.environ.items():
+            if k.startswith(PREFIX):
+                self.log.debug('Seeing environ "%s"="%s"', k, v)
+                # use __ instead of . as separator in env variable.
+                # Warning, case sensitive !
+                _, *path, key = k.split("__")
+                section = new_config
+                for p in path:
+                    section = section[p]
+                setattr(section, key, DeferredConfigString(v))
```

Comment:
We do not handle errors her, I'm not sure if we should keep going on invalid values.

Reply:
I think it should error out if an env variable is set to an invalid value.

Reply:
I tested and we do actually already error in the same way than when we pass an invalid CLI value (`MYAPP__Foo__i=a42` or `--Foo.i=a42`)


<details><summary>TB when invalid value</summary>
<code>

```
Traceback (most recent call last):
  File "/Users/bussonniermatthias/dev/traitlets/examples/myapp.py", line 155, in <module>
    main()
  File "/Users/bussonniermatthias/dev/traitlets/examples/myapp.py", line 150, in main
    app.initialize()
  File "/Users/bussonniermatthias/dev/traitlets/examples/myapp.py", line 129, in initialize
    self.init_foo()
  File "/Users/bussonniermatthias/dev/traitlets/examples/myapp.py", line 118, in init_foo
    self.foo = Foo(parent=self)
  File "/Users/bussonniermatthias/dev/traitlets/examples/myapp.py", line 56, in __init__
    super().__init__(**kwargs)
  File "/Users/bussonniermatthias/dev/traitlets/traitlets/config/configurable.py", line 108, in __init__
    self.config = config
  File "/Users/bussonniermatthias/dev/traitlets/traitlets/traitlets.py", line 688, in __set__
    self.set(obj, value)
  File "/Users/bussonniermatthias/dev/traitlets/traitlets/traitlets.py", line 677, in set
    obj._notify_trait(self.name, old_value, new_value)
  File "/Users/bussonniermatthias/dev/traitlets/traitlets/traitlets.py", line 1459, in _notify_trait
    self.notify_change(
  File "/Users/bussonniermatthias/dev/traitlets/traitlets/traitlets.py", line 1471, in notify_change
    return self._notify_observers(change)
  File "/Users/bussonniermatthias/dev/traitlets/traitlets/traitlets.py", line 1518, in _notify_observers
    c(event)
  File "/Users/bussonniermatthias/dev/traitlets/traitlets/traitlets.py", line 1100, in compatible_observer
    return func(self, change)
  File "/Users/bussonniermatthias/dev/traitlets/traitlets/config/configurable.py", line 218, in _config_changed
    self._load_config(change.new, traits=traits, section_names=section_names)
  File "/Users/bussonniermatthias/dev/traitlets/traitlets/config/configurable.py", line 180, in _load_config
    setattr(self, name, deepcopy(config_value))
  File "/Users/bussonniermatthias/dev/traitlets/traitlets/traitlets.py", line 688, in __set__
    self.set(obj, value)
  File "/Users/bussonniermatthias/dev/traitlets/traitlets/traitlets.py", line 662, in set
    new_value = self._validate(obj, value)
  File "/Users/bussonniermatthias/dev/traitlets/traitlets/traitlets.py", line 694, in _validate
    value = self.validate(obj, value)
  File "/Users/bussonniermatthias/dev/traitlets/traitlets/traitlets.py", line 2341, in validate
    self.error(obj, value)
  File "/Users/bussonniermatthias/dev/traitlets/traitlets/traitlets.py", line 798, in error
    raise TraitError(e)
traitlets.traitlets.TraitError: The 'i' trait of a Foo instance expected an int, not the str 'a42'.
```

</code>
</details>

Reply:
Yeah, you don't want to pass some security-related option and have it ignored.